File: array_and_hashing/group_anagrams.py
from collections import defaultdict


def group_anagrams_sort(strs: list[str]) -> list[list[str]]:
    ans = defaultdict(list)

    for string in strs:
        key = ''.join(sorted(string))
        ans[key].append(string)

    return list(ans.values())


def group_anagrams_ascii(strs: list[str]) -> list[list[str]]:
    ans = defaultdict(list)

    for string in strs:
        counter = [0] * 26
        for char in string:
            counter[ord(char) - ord('a')] += 1
        ans[tuple(counter)].append(string)

    return list(ans.values())

File: array_and_hashing/group_anagrams.ts
function groupAnagramsSort(strs: string[]): string[][] {
    const ans = new Map<string, string[]>();

    for (const str of strs) {
        const key = str.split('').sort().join('');
        if (!ans.has(key)) ans.set(key, []);
        ans.get(key)!.push(str);
    }

    return [...ans.values()];
}

function groupAnagramsAscii(strs: string[]): string[][] {
    const ans = new Map<string, string[]>();

    for (const str of strs) {
        const counter = new Array(26).fill(0);
        for (const char of str) counter[char.charCodeAt(0) - 97]++;
        const key = counter.join(',');
        if (!ans.has(key)) ans.set(key, []);
        ans.get(key)!.push(str);
    }

    return [...ans.values()];
}
